src/bin/ambit/cmd.rs

```diff
@@ -52,24 +53,122 @@ fn is_symlinked(link_name: &Path, target: &Path) -> bool {
+    let contains_pattern_char = |s: &str| -> bool { s.contains('*') || s.contains('?') };
+    for entry in spec.into_iter() {
+        if contains_pattern_char(&entry) {
+            fn get_matches(
+                ancestor_path: &Path,
+                expected_path_kind: &AmbitPathKind,
+                does_match: impl Fn(&str) -> bool,
+            ) -> AmbitResult<Vec<PathBuf>> {
+                let mut matches = Vec::new();
+                // Iterate through all entries in ancestor_path.
+                for path in fs::read_dir(ancestor_path)? {
+                    let path = path?.path();
+                    // Validify the current path.
+                    if let Some(file_name) = path.file_name() {
+                        if match expected_path_kind {
+                            AmbitPathKind::File => path.is_file(),
+                            AmbitPathKind::Directory => path.is_dir(),
+                        } && does_match(&file_name.to_string_lossy().to_string())
+                        {
+                            matches.push(path);
+                        }
+                    }
+                }
+                Ok(matches)
+            }
+            // The only valid path at the start is the starting path.
+            let mut valid_paths: Vec<PathBuf> = vec![start_path.clone()];
+            let components: Vec<String> = Path::new(&entry)
+                .components()
+                .map(|comp| comp.as_os_str().to_string_lossy().to_string())
+                .collect();
+            for (i, component) in components.iter().enumerate() {
+                let mut new_valid_paths: Vec<PathBuf> = Vec::new();
+                let expected_path_kind = if i < components.len() - 1 {
+                    // There are still more components to go, expect a directory.
+                    AmbitPathKind::Directory
+                } else {
+                    // No more components, expect a file.
+                    AmbitPathKind::File
+                };
+                let pattern = Pattern::compile(
+                    &component,
+                    MatchOptions::WILDCARDS | MatchOptions::UNKNOWN_CHARS,
+                );
+                for path in &valid_paths {
+                    new_valid_paths.append(&mut get_matches(
+                        &path,
+                        &expected_path_kind,
+                        |name: &str| -> bool { pattern.matches(name) },
+                    )?);
+                }
+                valid_paths = new_valid_paths;
+            }
+            // Strip prefix from all paths.
+            valid_paths = valid_paths
+                .iter()
+                .map(|path| -> AmbitResult<PathBuf> {
+                    Ok(path.strip_prefix(&start_path)?.to_path_buf())
+                })
+                .collect::<AmbitResult<Vec<PathBuf>>>()?;
+            paths.append(&mut valid_paths);
+        } else {
+            paths.push(PathBuf::from(&entry));
```

Comment:
You may want to move the simpler case above, so the if statement turns into `if !contains_pattern_char(&entry) {`. It is also useful to add a comment why this is a simpler case.